x-pack/elastic-agent/pkg/agent/program/spec.go

```diff
@@ -39,6 +39,7 @@ type Spec struct {
 	PostInstallSteps  *transpiler.StepList `yaml:"post_install"`
 	PreUninstallSteps *transpiler.StepList `yaml:"pre_uninstall"`
 	When              string               `yaml:"when"`
+	Constraints       string               `yaml:"constraints"`
```

Comment:
i added contraints section which is basically a as when section eql expression, evaluated before transformations to save some time.
runtime.os/arch were injected into a tree beforehand, so far as a separate keys not a map, we need to add support for accessing dict elements in eql expression 
cc @blakerouse as you already approved the PR

Reply:
Small question, we are adding a new key "constraints", could this be handled already by the "when" key?

Reply:
it could but it would mean we do the transformation which are not even necessary and then filter out all of the results because precondition failed. i wanted us to spare compute cycles